### ISSUE TITLE
kodiak is not closing issues through keyword usage 

### ISSUE THREAD

**Issue by tobilen** [chdsbd/kodiak#338]
possibly a duplicate of https://github.com/chdsbd/kodiak/issues/122

We've been using kodiak for a couple of months without problems up until a couple of days ago, when issue closing within our org stopped working. Merging still works fine, but the automation does not trigger.

Key facts that might be important:
- Issue and PR are in different repositories
- Issue is linked in multiple projects
- kodiak has permission in both repositories to read/write issues

What we've tried:
- Removing and readding kodiak integration in the org management

Let me know if you need the org name or any other details.

**Comment by chdsbd** [comment]
Hi @tobilen,

Thanks for opening this issue. I think I may have just reproduced this.

It seems that only cross-repository issue closing is affected. If I try to close a cross-repository issue, the GitHub UI looks like it will work, but when the PR is merged, nothing happens.

If `closes https://github.com/chdsbd/other_repo_test/issues/1` ends up in the merge commit body, GitHub will close the related issue via that commit message, but the PR itself doesn't close the issue like it would normally.

I believe this is a bug with GitHub.

As a workaround, if you configure Kodiak in such a way that `closes https://github.com/chdsbd/other_repo_test/issues/1` ends up in the merge commit, I think GitHub will close the issue because of the commit message.

I'm going to open a support ticket with GitHub about this.

**Comment by tobilen** [comment]
Thanks so much for the quick and well researched answer. Seems like you've pinpointed the issue.

I'll try writing in the merge commit, but this is probably not something that can be easily disseminated  in a bigger group of developers who also work on other repositories.
If you can provide me with a link to the issue you're planing on opening, you can feel free to close this one.

We tried merging by hand before i opened this issue and closing worked fine, so i can confirm this finding as well.

**Comment by chdsbd** [comment]
Kodiak can use the PR body as the merge commit message via [`merge.message.body`](https://kodiakhq.com/docs/config-reference#mergemessagebody). This way when you write `closes https://github.com/chdsbd/other_repo_test/issues/1` in the PR body Kodiak will add it to the merge commit message.

I submitted a support request with GitHub, but there isn't a way to link to that so I'll post any replies I receive from GitHub in this thread.

**Comment by tobilen** [comment]
i've tried that, but it seems the automation is not working like you described. this is my merge commit:
![image](https://user-images.githubusercontent.com/8217788/79582604-48209500-80cc-11ea-8e85-acbb5632800f.png)

it's an empty PR (as you can see by the changes), but i don't think that makes a difference.

**Comment by chdsbd** [comment]
Okay, this is pretty weird. It seems like it only works with "squash" (maybe rebase two, haven't tested). Using "merge" doesn't work.

GitHub also renders the commit messages differently for some reason.

"squash" merge method
<img width="952" alt="image" src="https://user-images.githubusercontent.com/1929960/79588118-763caf80-80a1-11ea-8fd4-32cb41594387.png">
"merge" merge method
<img width="952" alt="image" src="https://user-images.githubusercontent.com/1929960/79588159-82c10800-80a1-11ea-9444-e7c7eaa17f9e.png">

**Comment by tobilen** [comment]
the difference in how the issue links are presented might stem from copypasting the PR body. This happened to me as well.

But yes, this seems dependent on merge method. You also have the dotted line under the `closes` in the squash PR

**Comment by chdsbd** [comment]
I think there is a second bug where GitHub is not rendering the `closes https://github.com/chdsbd/other_repo_test/issues/2` correctly when the merge commit is created via the GitHub API. If I merge the PR and specify the commit message through the UI, GitHub will render the merge commit with the dotted line under the `close`.

I'm going to make another ticket with GitHub support.

**Comment by chdsbd** [comment]
still waiting on a response from GitHub Support.

**Comment by chdsbd** [comment]
GitHub responded! I've pasted the responses below for posterity and will update this ticket when I hear more.

### cross-repository `closes` keyword not working


> Christopher Dignam
> Apr 17, 12:44 PM UTC
> 
> Hello,
> A new bug has appeared for GitHub Apps.
> 
> For a pull request that references a cross-repository issue in the PR body like `closes https://github.com/acme/my_other_repo/issues/254`, GitHub says it will close the linked PR on merge, but when the GitHub App merges the PR, the linked issue is _not_ closed like GitHub says it would.
> 
> If a regular GitHub user merges that PR instead of a GItHub App, the linked PR does get closed as intended.
> 
> There's a public issue for this bug here: https://github.com/chdsbd/kodiak/issues/338
> 
> Let me know if you need any more information.
>
> ---
>
> *Redacted* (GitHub Support)
> May 6, 7:20 AM UTC
> 
> Hi Christopher,
> 
> Thanks for providing all that information. We have been able to reproduce what you are seeing and have reached out to our internal engineering teams for more information. While I cannot give an ETA, I'll update you once I have more information to share.
> 
> Cheers,
> *Redacted*


### incorrect commit rendering through GitHub API

> Christopher Dignam
> Apr 17, 4:45 PM UTC
> 
> When I merge a PR through [the GitHub API](https://developer.github.com/v3/pulls/#merge-a-pull-request-merge-button), if I choose the "merge" merge_method, my commit message will not be formatted correctly with the [`closes` keyword](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).
> 
> If I merge the PR using the GitHub UI with the same content, the resulting merge commit message is rendered correctly with the dotted underline on the `closes` keyword.
> 
> If I choose the "squash" merge_method in the API, the resulting merge commit body gets rendered correctly.
> 
> This seems to be a specific issue with this API endpoint for the "merge" merge_method.
> 
> This rendering issue is also not just a UI issue. For the merge commit created by the GitHub API, the keywords do not trigger the closing of linked issues.
> 
> The first attached image shows a merge commit created via the API, with the keywords not rendering correctly and the associated issue not being closed.
> The second attached image shows a merge commit created via the UI, with the keywords rendered correctly and the associated PR being closed as expected.
> 
> Let me know if you need any more information.
> 
> Attachment(s)
> [79588159-82c10800-80a1-11ea-9444-e7c7eaa17f9e.png](https://enterprise.githubsupport.com/attachments/token/kZnWyKNkFMnu6vvhjwYa8Tkfz/?name=79588159-82c10800-80a1-11ea-9444-e7c7eaa17f9e.png)
> [79588118-763caf80-80a1-11ea-8fd4-32cb41594387.png](https://enterprise.githubsupport.com/attachments/token/Xhn7pjJLEaNmnkpw119FVtVhU/?name=79588118-763caf80-80a1-11ea-8fd4-32cb41594387.png)
>
> ---
> 
> *Redacted* (GitHub Support)
> May 6, 7:32 AM UTC
> 
> Hi Christopher,
> 
> Apologies for the time it took us to get back to you. We recently saw some of our support request queues overflow and we are slowly regaining control of it.
> 
> I also just worked on your other support request regarding GitHub Apps not being able to close cross-repository issues and I wonder whether both behaviors are somehow related, i.e. share a common root cause.
> 
> I'll put this ticket on hold while we investigate further and will update you as soon as I have news to share.
> 
> Thanks,
> *Redacted*

**Comment by tobilen** [comment]
thanks for sharing those 👍

**Comment by beckerei** [comment]
As of today all of the referenced issues get closed automatically 👍

**Comment by tobilen** [comment]
yes, i'll close this since it seems to be fixed
